### PR TITLE
Color composition

### PR DESCRIPTION
**Pull request by stephencelis**
Normalizes the color type and, using some isomorphic functions, allows us to transform colors based on hue, saturation, lightness, etc.

### CODE REVIEW

**Comment by stephencelis** [comment]
- [ ] Figure out pretty-printing (`rgb($1, $2, $3, $4)` vs `rgb($1,$2,$3,$4)` vs `#\h{6}`)

**Comment by stephencelis** [comment]
- [ ] Figure out pretty-printing (`rgba($1, $2, $3, $4)` vs `rgba($1,$2,$3,$4)` vs `#\h{6}`, etc.)

**Comment by stephencelis** [comment]
Also: may need to file some Swift bugs around key-paths and computed properties...

File: Sources/CSS/Color.swift [hunk 1]
```diff
@@ -1,66 +1,219 @@
-public enum Color {
-  case rgba(Int, Int, Int, Float)
-  case hsla(Int, Float, Float, Float)
-  case custom(Value)
+import Prelude
+
+public struct Color {
+  private(set) var redComponent: Float
```
Comment (mbrandonw) [inline Sources/CSS/Color.swift]: curious if it's better to store these as `UInt8` instead? all of the initializers/getters/setters could still be `Int`s

the only thing that made me think of it was seeing the definition of `==` down below and remembering that's not good to compare equality of floats.

File: Sources/CSS/Color.swift [hunk 2]
```diff
@@ -1,66 +1,219 @@
...
-    case let (.custom(lhs), .custom(rhs)):
-      return lhs == rhs
-    case (.rgba, .hsla), (.rgba, .custom), (.hsla, .rgba), (.hsla, .custom), (.custom, .rgba),
-         (.custom, .hsla):
-      return false
-    }
+    return lhs.red == rhs.red
+      && lhs.blue == rhs.blue
+      && lhs.green == rhs.green
+      && lhs.alpha == rhs.alpha
```
Comment (mbrandonw) [inline Sources/CSS/Color.swift]: or another approach would be to convert to `Int`s here and compare them

File: Sources/CSS/Color.swift [hunk 3]
```diff
@@ -1,66 +1,219 @@
...
-      return .init(.plain("rgba(\(r), \(g), \(b), \(a))"))
-    case let .hsla(h, s, l, 1):
-      return .init(.plain("hsla(\(h), \(s), \(l))"))
-    case let .hsla(h, s, l, a):
-      return .init(.plain("hsla(\(h), \(s), \(l), \(a))"))
-    case let .custom(str):
-      return str
-    }
+    return .init(
+      .plain("rgba(\(toUInt8(self.red)), \(toUInt8(self.green)), \(toUInt8(self.blue)), \(self.alpha))")
```
Comment (mbrandonw) [inline Sources/CSS/Color.swift]: checking for `alpha == 1` and omitting that from the CSS would save us some bytes (lol... half joking)

File: Sources/CSS/Color.swift [hunk 4]
```diff
@@ -1,66 +1,219 @@
...
+    )
+  }
+}
+
+private func toUInt8(_ x: Float) -> Int {
+  return Int(max(0, min(255, x * 255)).rounded())
+}
+
+public func darken(_ by: Float) -> (Color) -> Color {
+  return over(\Color.lightness) <| { $0 * (1 - by) }
```
Comment (mbrandonw) [inline Sources/CSS/Color.swift]: fun!

File: Sources/CSS/Color.swift [hunk 5]
```diff
@@ -1,66 +1,219 @@
...
+    (r1, g1, b1) = (0, 0, 0)
+  }
+
+  let m = l - 0.5 * c
+  let (r, g, b) = (r1 + m, g1 + m, b1 + m)
+
+  return (r, g, b)
+}
+
+fileprivate func clamped<T>(_ to: Range<T>) -> (T) -> T {
```
Comment (mbrandonw) [inline Sources/CSS/Color.swift]: good function for our prelude someday. I had one similar in the ks prelude

File: Sources/CSS/Color.swift [hunk 6]
```diff
@@ -1,66 +1,219 @@
...
+  private(set) var blueComponent: Float
+  private(set) var greenComponent: Float
+  private(set) var alphaComponent: Float
+
+  public private(set) var red: Float {
+    get {
+      return self.redComponent
+    }
+    set(r) {
+      self.redComponent = (clamped(0..<1) <| r)
```
Comment (mbrandonw) [inline Sources/CSS/Color.swift]: are these extra parens needed?

File: Sources/CSS/Color.swift (hunk 3 above)
Comment (mbrandonw) [inline Sources/CSS/Color.swift]: Oh sorry didn't see your previous comment

File: Sources/CSS/Color.swift (hunk 1 above)
Comment (stephencelis) [inline Sources/CSS/Color.swift]: Maybe! It'd be a little lossy, but maybe good for our purposes here.

File: Sources/CSS/Color.swift (hunk 6 above)
Comment (stephencelis) [inline Sources/CSS/Color.swift]: Yeah. Assignment precedence :/ Should fix in Prelude.

File: Sources/CSS/Color.swift (hunk 3 above)
Comment (stephencelis) [inline Sources/CSS/Color.swift]: Don't joke! We should def do this!

**Pull request merged**